on export: added a second overwrite option (overwrite without confirm…

bertwim opened the pull request:
…ation)

This options facilitates workflows where it is known in advance that the exported image can always overwrite an existing image with the same name.

elstoc commented:
IMO both of these options (to prompt or not to prompt) miss the point. Popping up a single warning before a batch of exports saying "you might overwrite some stuff" is not a good solution, but neither is _not_ popping up that warning. It assumes you know exactly what images are in the target directory and are sure you want to do the same thing to all of them.

There's a standard way to deal with this sort of thing and it's to pop up a message for each conflicting file and give options, along with the additional option to "do the same thing for all other conflicts", as I've suggested in #3816. I agree that the existing prompt is annoying and doesn't really do anything useful, but I'd rather fix it properly than just have an option to get rid of the prompt again.

Presumably we still need to make any solution work for darktable-cli as well.

On a separate point, you've changed what the value of "2" means in `dt_disk_onconflict_actions_t`. This means that when a user upgrades, if their previous export setting was "skip", you've just updated it to "overwrite without prompting". This needs to be fixed.

bertwim commented:
Thanks for responding,

On the overwrite:
in my workflow I *do* know that I want to overwrite on conflict. I work 
from raw files;  any jpg (or whatever the output format is) in the 
target directory I consider as being generated by DT from these raw 
files. Worst case scenario is that a file is unnecessarily exported, 
thus overwriting an identical copy. This workflow is, of course, a 
choice, but it does yield a simple and convenient one, which easily 
lends itself for automation. That is why I need this option, not 
replacing the other overwrite, but as an additional one.

On the index 2: you are quite right. I changed it only to have a more 
natural order in the menu. Now, on the first startup DT will preset the 
export-option indeed to a different choice, which you then have to reset 
manually. After that, all will be ok again.
It is easy to change this, of course. But what is the proper way here? 
It seems strange that the order in which the menu items appear is 
related to the numeric value of the enumeration. Maybe I just 
misunderstood how this all works?

Regards
Bertwim


On 3/11/21 4:11 PM, Chris Elston wrote:
>
> IMO both of these options (to prompt or not to prompt) miss the point. 
> Popping up a single warning before a batch of exports saying "you 
> might overwrite some stuff" is not a good solution, but neither is 
> /not/ popping up that warning. It assumes you know exactly what images 
> are in the target directory and are sure you want to do the same thing 
> to all of them.
>
> There's a standard way to deal with this sort of thing and it's to pop 
> up a message for each conflicting file and give options, along with 
> the additional option to "do the same thing for all other conflicts", 
> as I've suggested in #3816 
> <https://github.com/darktable-org/darktable/issues/3816>. I agree that 
> the existing prompt is annoying and doesn't really do anything useful, 
> but I'd rather fix it properly than just have an option to get rid of 
> the prompt again.
>
> Presumably we still need to make any solution work for darktable-cli 
> as well.
>
> On a separate point, you've changed what the value of "2" means in 
> |dt_disk_onconflict_actions_t|. This means that when a user upgrades, 
> if their previous export setting was "skip", you've just updated it to 
> "overwrite without prompting". This needs to be fixed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/darktable-org/darktable/pull/8462#issuecomment-796807351>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AABDMW2F5EMEJOFLHXPLC3DTDDMT3ANCNFSM4ZAMX2LA>.
>

quovadit commented:
I agree with @elstoc that only prompting on actual conflicts is the better solution.
But if this new option 'ask what to do' is added, the prompt could be removed from the existing 'overwrite' option.
This way both requests are fulfilled.

bertwim commented:
Why are these Win64 checks still failing?

I think i have addressed the comments made on my contribution. The changes in disk.c compared with the original version are three trivial changes, thus preserving compatibility, as was a valid comment made by @elstoc.

Whereas the comments made about prompting etc. are all valid remark for certain workflows, I think I explained above that for the workflow I follow,  a promptless overwrite is the only practical solution. 

Kind Regards
Bertwim

quovadit commented:
I think it's ok that overwrite is now promtless.
But if there is a separate option: 'on conflict: prompt', then it should only prompt if there actually is a conflict.

elstoc commented on src/imageio/storage/disk.c:
```diff
@@ -201,8 +202,9 @@ void gui_init(dt_imageio_module_storage_t *self)
   dt_bauhaus_widget_set_label(d->onsave_action, NULL, N_("on conflict"));
   dt_bauhaus_combobox_add(d->onsave_action, _("create unique filename"));
-  dt_bauhaus_combobox_add(d->onsave_action, _("overwrite"));
+  dt_bauhaus_combobox_add(d->onsave_action, _("prompt"));
```
"prompt" isn't descriptive enough - it doesn't say what it actually does - and there's no tooltip to provide a clue.

elstoc commented:
Thinking about it, it might be better to keep the same set of options that we have today but add an extra preference to the 'security' tab of the preferences window. Something like "ask before exporting in overwrite mode".

quovadit commented:
For me it would be perfect to call it 'ask what to do' and offer options in the popup for every conflict:
- rename
- skip
- overwrite

... with a checkbox 'apply to all conflicts'

elstoc commented:
@quovadit I agree, but that's a lot more complicated a change than this one (which is why I haven't done it yet). For now I think it's a question of do we want to merge this PR, or is there a way we could make it better. I think the security preference would make it less intrusive but don't know whether it's better or not.

quovadit commented:
If we only want to have an option not to ask, then I would not add a 4th option 'prompt' but a preference in the tab security 'ask before overwriting files on export' with default true.
